packages/server/src/fhir/sql.ts

```diff
@@ -396,17 +396,25 @@ export class SqlBuilder {
 
       return { rowCount: result.rowCount ?? 0, rows: result.rows };
     } catch (err: any) {
-      if (err && typeof err === 'object' && err.code === '23505') {
-        // Catch duplicate key errors and throw a 409 Conflict
-        // See https://github.com/brianc/node-postgres/issues/1602
-        // See https://www.postgresql.org/docs/10/errcodes-appendix.html
-        throw new OperationOutcomeError(conflict(err.detail));
-      }
-      throw err;
+      throw normalizeDatabaseError(err);
     }
   }
 }
 
+export function normalizeDatabaseError(err: any): OperationOutcomeError {
+  if (err?.code === '23505') {
+    // Catch duplicate key errors and throw a 409 Conflict
+    // See https://github.com/brianc/node-postgres/issues/1602
+    // See https://www.postgresql.org/docs/10/errcodes-appendix.html
+    return new OperationOutcomeError(conflict(err.detail));
+  }
+  if (err?.code === '40001') {
+    // Catch transaction serialization errors and throw a 409 Conflict
+    return new OperationOutcomeError(conflict(err.message));
```

Comment:
Should these 2 special cases be combined?  Not sure what the difference is between `err.message` and `err.detail`?

Reply:
It's mostly a choice of which has the more useful information: for error `23505`, `err.detail` contains the specific conflict key/value, but for error `40001` it would be a relatively arbitrary SQL statement from the transaction in conflict